### PR TITLE
feat: default variants

### PR DESCRIPTION
**Pull request by markphelps**
Redo of #3168 

Fixes: #3134

CleanShot 2024-07-19 at 13 37 59@2x

CleanShot 2024-07-19 at 13 50 07@2x

CleanShot 2024-07-19 at 13 51 47@2x
CleanShot 2024-07-19 at 13 51 58@2x

## TODO (In this PR)

- [x] Add tests
- [x] Implement in UI
- [x] Add for other relational dbs
- [x] Import/Export support (will also influence declarative backends)
- [x] Add to snapshot builder code (declarative)

## In other repos

- [ ] Add to API docs
- [x] Add to client-side eval logic (https://github.com/flipt-io/flipt-client-sdks/pull/309)
- [ ] Update OpenFeature providers

### CODE REVIEW

**Comment by markphelps** [comment]
I'm opening this up for feedback before I get too far in the weeds with adding tests

cc @erka

**Comment by erka** [comment]
> I'm opening this up for feedback before I get too far in the weeds with adding tests
> 
> cc @erka

@markphelps  Overall, it looks fantastic. There is one white spot with `evaluation.proto`. I don't see that we set a default variant there.

**Comment by markphelps** [comment]
> > I'm opening this up for feedback before I get too far in the weeds with adding tests
> > cc @erka
> 
> @markphelps Overall, it looks fantastic. There is one white spot with `evaluation.proto`. I don't see that we set a default variant there.

Im not quite sure what you mean. Would you mind showing me in the a comment in the PR if possible?

**Comment by erka** [comment]
> Im not quite sure what you mean. Would you mind showing me in the a comment in the PR if possible?

If I set the default variant for a flag and then look at snapshot at http://localhost:8080/internal/v1/evaluation/snapshot/namespace/default I don't see any changes in the response.

**Comment by markphelps** [comment]
> > Im not quite sure what you mean. Would you mind showing me in the a comment in the PR if possible?
> 
> If I set the default variant for a flag and then look at snapshot at http://localhost:8080/internal/v1/evaluation/snapshot/namespace/default I don't see any changes in the response.

ah yes! good catch. will add. and also need to add support in the client side evaluator engine

**Comment by markphelps** [comment]
## Eval Data API

### With Proper Header Set

```
curl -H "X-Flipt-Accept-Server-Version: 1.47.0" http://localhost:8080/internal/v1/evaluation/snapshot/namespace/default | jq

{
  "namespace": {
    "key": "default"
  },
  "flags": [
    {
      "key": "fgsdaf",
      "name": "fgsdaf",
      "description": "",
      "enabled": false,
      "type": "BOOLEAN_FLAG_TYPE",
      "createdAt": "2024-07-19T14:18:19.835699Z",
      "updatedAt": "2024-07-19T14:18:19.835699Z",
      "rules": [],
      "rollouts": [
        {
          "type": "THRESHOLD_ROLLOUT_TYPE",
          "rank": 1,
          "threshold": {
            "percentage": 50,
            "value": true
          }
        }
      ]
    },
    {
      "key": "dfsdf",
      "name": "dfsdf",
      "description": "",
      "enabled": true,
      "type": "VARIANT_FLAG_TYPE",
      "createdAt": "2024-07-19T14:20:48.359953Z",
      "updatedAt": "2024-07-21T16:59:59.373240Z",
      "rules": [],
      "rollouts": [],
      "defaultVariant": {
        "id": "e468f221-7e48-4dcf-8f1f-3156a990f363",
        "key": "goo",
        "attachment": ""
      }
    }
  ]
}
```

---

### With Lower Version Header Set

```
curl -H "X-Flipt-Accept-Server-Version: 1.46.0" http://localhost:8080/internal/v1/evaluation/snapshot/namespace/default | jq

{
  "namespace": {
    "key": "default"
  },
  "flags": [
    {
      "key": "fgsdaf",
      "name": "fgsdaf",
      "description": "",
      "enabled": false,
      "type": "BOOLEAN_FLAG_TYPE",
      "createdAt": "2024-07-19T14:18:19.835699Z",
      "updatedAt": "2024-07-19T14:18:19.835699Z",
      "rules": [],
      "rollouts": [
        {
          "type": "THRESHOLD_ROLLOUT_TYPE",
          "rank": 1,
          "threshold": {
            "percentage": 50,
            "value": true
          }
        }
      ]
    },
    {
      "key": "dfsdf",
      "name": "dfsdf",
      "description": "",
      "enabled": true,
      "type": "VARIANT_FLAG_TYPE",
      "createdAt": "2024-07-19T14:20:48.359953Z",
      "updatedAt": "2024-07-21T16:59:59.373240Z",
      "rules": [],
      "rollouts": []
    }
  ]
}
```

---

### With No Header Set

```
curl http://localhost:8080/internal/v1/evaluation/snapshot/namespace/default | jq

{
  "namespace": {
    "key": "default"
  },
  "flags": [
    {
      "key": "fgsdaf",
      "name": "fgsdaf",
      "description": "",
      "enabled": false,
      "type": "BOOLEAN_FLAG_TYPE",
      "createdAt": "2024-07-19T14:18:19.835699Z",
      "updatedAt": "2024-07-19T14:18:19.835699Z",
      "rules": [],
      "rollouts": [
        {
          "type": "THRESHOLD_ROLLOUT_TYPE",
          "rank": 1,
          "threshold": {
            "percentage": 50,
            "value": true
          }
        }
      ]
    },
    {
      "key": "dfsdf",
      "name": "dfsdf",
      "description": "",
      "enabled": true,
      "type": "VARIANT_FLAG_TYPE",
      "createdAt": "2024-07-19T14:20:48.359953Z",
      "updatedAt": "2024-07-21T16:59:59.373240Z",
      "rules": [],
      "rollouts": []
    }
  ]
}
```

File: internal/server/evaluation/data/server.go [hunk 1]
```diff
@@ -97,7 +97,26 @@ func toEvaluationRolloutType(r flipt.RolloutType) evaluation.EvaluationRolloutTy
...
+	return m.msg
+}
+
+var (
+	supportsEntityIdConstraintMinVersion = minVersion{
+		Version: semver.MustParse("1.38.0"),
+		msg:     "skipping `entity_id` constraint type to support older client; upgrade client to allow `entity_id` constraint type",
+	}
+
+	supportsDefaultVariantMinVersion = minVersion{
```
Comment (erka) [inline internal/server/evaluation/data/server.go:115]: Should we really worry about it? It probably will not break old clients.

**Review by erka — APPROVED** [review]
it looks great. I have only one comment about compatibility.

File: internal/server/evaluation/data/server.go (hunk 1 above)
Comment (markphelps) [reply internal/server/evaluation/data/server.go:115]: you're probably right, i just prefer to err on the side of caution

**Review by arussellsaw — APPROVED** [review]

**Pull request merged**